### DISCUSSION TITLE
Experience Cloud Guest User

### DISCUSSION THREAD

**Discussion by taits** [jongpie/NebulaLogger#209] [Q&A] [answered]
Wondering if logging from Apex executed with an Experience Cloud guest user should work?  Any specific permissions need to be added to the Sites Guest user Profile?

Seeing a Salesforce System Error on line 2656 of Logger.apxc
`String transactionId = System.Request.getCurrent().getRequestId();`

**Comment by jongpie** [comment, marked as answer]
Hey @taits! Yeah, my goal is that logging should work for Experience Cloud (including the guest user), but it sounds like there might be a bug for the guest user scenario. Any chance you could share a stack trace string for the error that happens on that line? My guess is that for the guest user, Salesforce probably doesn't provide a request ID, so I'm guessing `System.Request.getCurrent()` is null --> it fails on the chained call to `getRequestId()`. If that's the case, then I think it should be a (hopefully) easy fix to make.

**Reply by mohitlakde** [reply]
Hello @jongpie Please see stack trace. cc: @taits 

12:57:01.117 (266470405)|METHOD_ENTRY|[12]|01p0q000003OLey|Logger.Logger()
12:57:01.117 (266478981)|STATEMENT_EXECUTE|[12]
12:57:01.117 (266485482)|STATEMENT_EXECUTE|[12]
12:57:01.117 (266513908)|HEAP_ALLOCATE|[52]|Bytes:5
12:57:01.117 (266544080)|HEAP_ALLOCATE|[58]|Bytes:5
12:57:01.117 (266554247)|HEAP_ALLOCATE|[66]|Bytes:7
12:57:01.117 (266575570)|HEAP_ALLOCATE|[13]|Bytes:5
12:57:01.117 (266581475)|STATEMENT_EXECUTE|[13]
12:57:01.117 (266629137)|VARIABLE_ASSIGNMENT|[13]|Logger.DEFAULT_LOGGING_LEVEL|"DEBUG"|0xbe2202e
12:57:01.117 (266636831)|STATEMENT_EXECUTE|[14]
12:57:01.117 (266648865)|HEAP_ALLOCATE|[14]|Bytes:4
12:57:01.117 (266757180)|HEAP_ALLOCATE|[EXTERNAL]|Bytes:4
12:57:01.117 (266777535)|VARIABLE_ASSIGNMENT|[14]|Logger.LOG_ENTRIES_BUFFER|[]|0x7abf6672
12:57:01.117 (266785092)|STATEMENT_EXECUTE|[15]
12:57:01.117 (266810022)|METHOD_ENTRY|[15]|01p0q000003OLey|Logger.setTransactionId()
12:57:01.117 (266841342)|HEAP_ALLOCATE|[2656]|Bytes:5
12:57:01.117 (266846966)|STATEMENT_EXECUTE|[2655]
12:57:01.117 (266849226)|STATEMENT_EXECUTE|[2656]
12:57:01.117 (266879537)|VARIABLE_SCOPE_BEGIN|[17]|System.Request.instance|System.RequestImpl|true|true
12:57:01.117 (266892502)|HEAP_ALLOCATE|[2656]|Bytes:2
12:57:01.117 (266922110)|SYSTEM_METHOD_ENTRY|[16]|Request.Request()
12:57:01.117 (266927184)|STATEMENT_EXECUTE|[16]
12:57:01.117 (266943922)|SYSTEM_METHOD_EXIT|[16]|Request
12:57:01.117 (266955407)|METHOD_ENTRY|[2656]||System.Request.getCurrent()
12:57:01.117 (273549460)|METHOD_EXIT|[2656]||System.Request.getCurrent()
12:57:01.117 (273573260)|METHOD_EXIT|[15]|01p0q000003OLey|Logger.setTransactionId()
12:57:01.117 (273587303)|METHOD_EXIT|[12]|Logger
12:57:01.117 (274439434)|FATAL_ERROR|System.UnexpectedException: Salesforce System Error: 2059952125-1101361 (-334555042) (-334555042)

Class.System.RequestImpl.<init>: line 26, column 1
Class.System.Request.getCurrent: line 34, column 1
Class.Logger.setTransactionId: line 2656, column 1
Class.Logger: line 15, column 1
Class.KDP_SocialRegistrationHandler.updateUser: line 63, column 1
12:57:01.275 (275107919)|CUMULATIVE_LIMIT_USAGE
12:57:01.275 (275107919)|LIMIT_USAGE_FOR_NS|(default)|
  Number of SOQL queries: 0 out of 100
  Number of query rows: 0 out of 50000
  Number of SOSL queries: 0 out of 20
  Number of DML statements: 0 out of 150
  Number of Publish Immediate DML: 0 out of 150
  Number of DML rows: 0 out of 10000
  Maximum CPU time: 0 out of 10000
  Maximum heap size: 0 out of 6000000
  Number of callouts: 0 out of 100
  Number of Email Invocations: 0 out of 10
  Number of future calls: 0 out of 50
  Number of queueable jobs added to the queue: 0 out of 50
  Number of Mobile Apex push calls: 0 out of 10

**Reply by jongpie** [reply]
@mohitlakde thanks for sharing the stack trace! I do have some unit tests for guest users, but the tests are passing, so I think this may be an issue that only happens when actually using an Experience site (not in a test context). I'll try to setup an Experience site in a scratch org & see if I can reproduce this issue - I'll let you and @taits what I find out!

**Reply by mohitlakde** [reply]
Thanks @jongpie. Exactly I also noticed it works in Test Class but not for actual user. Let me know your findings.

**Reply by jongpie** [reply]
Hey @mohitlakde - I've tried a few things in a scratch org, but so far, I have not been able to reproduce this issue with the guest user. I tested by creating an Experience Site with self-registration enabled, using a sample VF page & controller class for the self-registration (which runs as the site's guest user). Within the controller class, I added some logging, and it seems to work for me (see screenshot of the resulting log).

![image](https://user-images.githubusercontent.com/1267157/135009794-d150f0d2-9d28-46d6-9114-673d52c4a7ad.png)

Could you share some more details regarding your exact scenario? Based on the stack trace you provided, it looks like a registration class is running when the error occurs - any other info you can share would be helpful (including which version of Nebula Logger you're using). And are you able to reproduce the issue in one of your sandboxes/scratch orgs?

**Reply by mohitlakde** [reply]
Thank you @jongpie Sorry not to mentioned earlier, apart from Community normal login, user can login through Social platform as well. I have not tested with regular Community logged-in user, but can you test for Social(Facebook or Google) Logged-in user.

Scenario: 
Every time when User logs in through social platform we are capturing incoming details in Salesforce. 

Logger.info(String.valueOf(data)); //data is incoming Auth.UserData
Logger.saveLog();

**Reply by mohitlakde** [reply]
@jongpie i tried for regular community user as you mentioned and it worked, but still fails for community user when they login using social accounts, specifically if we put this logger saving line in social controller.

**Reply by jongpie** [reply]
@mohitlakde thanks for the details! I was finally able to reproduce this issue in a scratch org - I've logged it as a bug, issue #224. I was able to create a quick prototype to bypass the issue, and I was then able to successfully save logs from an auth provider registration class (see screenshot below), so there should definitely be a way to get this working soon. I'm going to keep exploring options for how to best implement a workaround for the problem - I'll try to have a fix done ASAP!

![image](https://user-images.githubusercontent.com/1267157/136727231-eea34485-7902-4523-b997-a5762696cafd.png)

**Reply by jongpie** [reply]
@mohitlakde I think I've found a workaround for this issue - could you try out this beta package version in one of your sandboxes/scratch orgs to see if it solves the issue for you? It solved it for me in a scratch org I setup, but just want to be sure it's working for you too, before I merge the changes.

- New beta package version - [https://test.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015l1OQAQ](https://test.salesforce.com/packaging/installPackage.apexp?p0=04t5Y0000015l1OQAQ)
- Relevant changes - https://github.com/jongpie/NebulaLogger/pull/218/commits/d1ca0558d276f8d35f6fad53530c9b758ec95b94

I'm hoping to include this in PR #218 (v4.6.10) - let me know if you run into any issues.

**Reply by jongpie** [reply]
@mohitlakde and @taits I did some more testing and everything seems to be working now, so I've included a bugfix for this in the new release [v4.6.10](https://github.com/jongpie/NebulaLogger/releases/tag/v4.6.10). Hopefully everything is now resolved, but let me know if you run into any more issues!
